File: midolmanj-mgmt/src/main/java/com/midokura/midolman/mgmt/host/rest_api/TunnelZoneHostResource.java
/*
 * Copyright 2012 Midokura PTE LTD.
 */
package com.midokura.midolman.mgmt.host.rest_api;

import com.midokura.midolman.mgmt.ResourceUriBuilder;
import com.midokura.midolman.mgmt.VendorMediaType;
import com.midokura.midolman.mgmt.auth.AuthRole;
import com.midokura.midolman.mgmt.host.GreTunnelZoneHost;
import com.midokura.midolman.mgmt.host.CapwapTunnelZoneHost;
import com.midokura.midolman.mgmt.host.TunnelZoneHost;
import com.midokura.midolman.mgmt.host.TunnelZoneHostFactory;
import com.midokura.midolman.mgmt.rest_api.BadRequestHttpException;
import com.midokura.midolman.mgmt.rest_api.NotFoundHttpException;
import com.midokura.midolman.state.StateAccessException;
import com.midokura.midonet.cluster.DataClient;
import com.midokura.midonet.cluster.data.TunnelZone;

import com.google.inject.Inject;
import com.google.inject.assistedinject.Assisted;
import com.google.inject.servlet.RequestScoped;
import javax.annotation.security.RolesAllowed;
import javax.validation.ConstraintViolation;
import javax.validation.Validator;
import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriInfo;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.UUID;

/**
 * REST API handler for tunnel zone - host mapping.
 */
@RequestScoped
public class TunnelZoneHostResource {

    private final UUID tunnelZoneId;
    private final UriInfo uriInfo;
    private final Validator validator;
    private final DataClient dataClient;

    @Inject
    public TunnelZoneHostResource(UriInfo uriInfo,
                                  Validator validator,
                                  DataClient dataClient,
                                  @Assisted UUID tunnelZoneId) {
        this.uriInfo = uriInfo;
        this.validator = validator;
        this.dataClient = dataClient;
        this.tunnelZoneId = tunnelZoneId;
    }

    private <T extends TunnelZoneHost> Response createTunnelZoneHost(T tzHost)
            throws StateAccessException {
        tzHost.setTunnelZoneId(tunnelZoneId);
        Set<ConstraintViolation<T>> violations =validator.validate(tzHost);
        if (!violations.isEmpty())
            throw new BadRequestHttpException(violations);

        dataClient.tunnelZonesAddMembership(tunnelZoneId,
                tzHost.toData());

        return Response.created(
                ResourceUriBuilder.getTunnelZoneHost(uriInfo.getBaseUri(),
                        tunnelZoneId, tzHost.getHostId()))
                .build();
    }

    @POST
    @RolesAllowed({ AuthRole.ADMIN })
    @Consumes({ VendorMediaType.APPLICATION_GRE_TUNNEL_ZONE_HOST_JSON })
    public Response create(GreTunnelZoneHost tunnelZoneHost)
            throws StateAccessException {

        return createTunnelZoneHost(tunnelZoneHost);
    }

    @POST
    @RolesAllowed({ AuthRole.ADMIN })
    @Consumes({ VendorMediaType.APPLICATION_CAPWAP_TUNNEL_ZONE_HOST_JSON })
    public Response create(CapwapTunnelZoneHost tunnelZoneHost)
            throws StateAccessException {

        return createTunnelZoneHost(tunnelZoneHost);
    }

    private List<TunnelZoneHost> listTunnelZoneHosts(
            Class<? extends TunnelZoneHost> clazz) throws StateAccessException {
        Set<TunnelZone.HostConfig<?,
                ?>> dataList =
                dataClient.tunnelZonesGetMemberships(tunnelZoneId);
        List<TunnelZoneHost> tunnelZoneHosts =
                new ArrayList<TunnelZoneHost>();

        for (TunnelZone.HostConfig data : dataList) {
            TunnelZoneHost tzh =
                    TunnelZoneHostFactory.createTunnelZoneHost(
                            tunnelZoneId, data);
            if (clazz == null || tzh.getClass().equals(clazz)) {
                tzh.setBaseUri(uriInfo.getBaseUri());
                tunnelZoneHosts.add(tzh);
            }
        }

        return tunnelZoneHosts;
    }

    @GET
    @RolesAllowed({ AuthRole.ADMIN })
    @Produces({ VendorMediaType
            .APPLICATION_TUNNEL_ZONE_HOST_COLLECTION_JSON })
    public List<TunnelZoneHost> listUntypedTunnelZoneHosts() throws
        StateAccessException {

        return listTunnelZoneHosts(null);
    }

    @GET
    @RolesAllowed({ AuthRole.ADMIN })
    @Produces({ VendorMediaType
            .APPLICATION_GRE_TUNNEL_ZONE_HOST_COLLECTION_JSON })
    public List<TunnelZoneHost> listGreTunnelZoneHosts() throws
            StateAccessException {

        return listTunnelZoneHosts(GreTunnelZoneHost.class);
    }

    @GET
    @RolesAllowed({ AuthRole.ADMIN })
    @Produces({ VendorMediaType
            .APPLICATION_CAPWAP_TUNNEL_ZONE_HOST_COLLECTION_JSON })
    public List<TunnelZoneHost> listCapwapTunnelZoneHosts() throws
            StateAccessException {

        return listTunnelZoneHosts(CapwapTunnelZoneHost.class);
    }

    private TunnelZoneHost getTunnelZoneHost(
            Class<? extends TunnelZoneHost> clazz,
            UUID hostId) throws StateAccessException {
        TunnelZone.HostConfig data = dataClient
                .tunnelZonesGetMembership(tunnelZoneId, hostId);
        if (data == null) {
            throw new NotFoundHttpException("The resource was not found");
        }

        TunnelZoneHost tzh = TunnelZoneHostFactory.createTunnelZoneHost(
                tunnelZoneId, data);
        if (clazz != null && !tzh.getClass().equals(clazz))
            throw new NotFoundHttpException("The resource was not found");

        tzh.setBaseUri(uriInfo.getBaseUri());
        return tzh;
    }

    @GET
    @RolesAllowed({ AuthRole.ADMIN })
    @Produces({ VendorMediaType.APPLICATION_TUNNEL_ZONE_HOST_JSON })
    @Path("/{hostId}")
    public TunnelZoneHost getUntypedTunnelZoneHost(
            @PathParam("hostId") UUID hostId) throws StateAccessException {
        return getTunnelZoneHost(null, hostId);
    }

    @GET
    @RolesAllowed({ AuthRole.ADMIN })
    @Produces({ VendorMediaType.APPLICATION_GRE_TUNNEL_ZONE_HOST_JSON })
    @Path("/{hostId}")
    public TunnelZoneHost getGreTunnelZoneHost(@PathParam("hostId")
                                               UUID hostId) throws
            StateAccessException {

        return getTunnelZoneHost(GreTunnelZoneHost.class, hostId);
    }

    @GET
    @RolesAllowed({ AuthRole.ADMIN })
    @Produces({ VendorMediaType.APPLICATION_CAPWAP_TUNNEL_ZONE_HOST_JSON })
    @Path("/{hostId}")
    public TunnelZoneHost getCapwapTunnelZoneHost(@PathParam("hostId")
                                               UUID hostId) throws
            StateAccessException {

        return getTunnelZoneHost(CapwapTunnelZoneHost.class, hostId);
    }

    @DELETE
    @RolesAllowed({ AuthRole.ADMIN })
    @Path("/{hostId}")
    public void delete(@PathParam("hostId") UUID hostId)
            throws StateAccessException {

        TunnelZone.HostConfig data = dataClient
                .tunnelZonesGetMembership(tunnelZoneId, hostId);
        if (data == null) {
            return;
        }

        dataClient.tunnelZonesDeleteMembership(tunnelZoneId, hostId);
    }
}

File: midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/resources/TunnelZoneHostResource.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.resources

import java.util.{List => JList, Set => JSet, UUID}
import javax.validation.ConstraintViolation
import javax.ws.rs._
import javax.ws.rs.core.Response
import javax.ws.rs.core.Response.Status

import scala.collection.JavaConversions._
import scala.collection.JavaConverters._

import com.google.inject.Inject
import com.google.inject.servlet.RequestScoped

import org.midonet.cluster.rest_api.models.{TunnelZone, TunnelZoneHost}
import org.midonet.cluster.rest_api.{BadRequestHttpException, NotFoundHttpException}
import org.midonet.cluster.services.rest_api.MidonetMediaTypes._
import org.midonet.cluster.services.rest_api.resources.MidonetResource.{OkNoContentResponse, ResourceContext}

@RequestScoped
class TunnelZoneHostResource @Inject()(tunnelZoneId: UUID,
                                       resContext: ResourceContext)
    extends MidonetResource[TunnelZoneHost](resContext) {

    @GET
    @Produces(Array(APPLICATION_TUNNEL_ZONE_HOST_JSON,
                    APPLICATION_GRE_TUNNEL_ZONE_HOST_JSON))
    @Path("{id}")
    override def get(@PathParam("id") id: String,
                     @HeaderParam("Accept") accept: String): TunnelZoneHost = {
        // Go via the resource so that it populates the TunnelZoneHosts inner
        // fields (tunnel zone id and uri)
        val hostId = UUID.fromString(id)
        new TunnelZoneResource(resContext)
            .get(tunnelZoneId.toString, APPLICATION_TUNNEL_ZONE_JSON)
            .tzHosts.find(_.hostId == hostId)
            .getOrElse(throw new NotFoundHttpException("Resource not found"))
    }

    @GET
    @Produces(Array(APPLICATION_TUNNEL_ZONE_HOST_COLLECTION_JSON,
                    APPLICATION_GRE_TUNNEL_ZONE_HOST_COLLECTION_JSON))
    override def list(@HeaderParam("Accept") accept: String)
    : JList[TunnelZoneHost] = {
        // Go via the resource so that it populates the TunnelZoneHosts inner
        // fields (tunnel zone id and uri)
        new TunnelZoneResource(resContext)
            .get(tunnelZoneId.toString, APPLICATION_TUNNEL_ZONE_JSON)
            .tzHosts
    }

    @POST
    @Consumes(Array(APPLICATION_GRE_TUNNEL_ZONE_HOST_JSON,
                    APPLICATION_TUNNEL_ZONE_HOST_JSON))
    override def create(tunnelZoneHost: TunnelZoneHost,
                        @HeaderParam("Content-Type") contentType: String)
    : Response = {

        val validator = resContext.validator
        val violations: JSet[ConstraintViolation[TunnelZoneHost]] =
            validator.validate(tunnelZoneHost)
        if (!violations.isEmpty) {
            throw new BadRequestHttpException(violations)
        }

        zkLock {
            val tunnelZone = getResource(classOf[TunnelZone], tunnelZoneId)
            if (tunnelZone.tzHosts.exists(_.hostId == tunnelZoneHost.hostId)) {
                Response.status(Status.CONFLICT).build()
            } else {
                tunnelZoneHost.create(tunnelZone.id)
                tunnelZoneHost.setBaseUri(resContext.uriInfo.getBaseUri)
                tunnelZone.tzHosts.add(tunnelZoneHost)
                tunnelZone.hostIds.add(tunnelZoneHost.hostId)
                updateResource(tunnelZone,
                               Response.created(tunnelZoneHost.getUri).build())
            }
        }
    }

    @DELETE
    @Path("{id}")
    override def delete(@PathParam("id") id: String): Response = zkLock {
        val hostId = UUID.fromString(id)
        val tunnelZone = getResource(classOf[TunnelZone], tunnelZoneId)
        tunnelZone.tzHosts.asScala
            .find(_.hostId == hostId)
            .map(tunnelZoneHost => {
                tunnelZone.tzHosts.remove(tunnelZoneHost)
                tunnelZone.hostIds.remove(tunnelZoneHost.hostId)
                updateResource(tunnelZone, OkNoContentResponse)
            })
            .getOrElse(Response.status(Status.NOT_FOUND).build())
    }

}
